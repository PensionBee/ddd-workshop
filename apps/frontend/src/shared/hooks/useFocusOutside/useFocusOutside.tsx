import { useEffect } from "react";

export const useFocusOutside = <E extends HTMLElement>(
  ref: React.RefObject<E>,
  callback: (event: FocusEvent) => void
) => {
  useEffect(() => {
    const listener = (event: FocusEvent) => {
      if (!ref.current) return;
      if (event.target && ref.current.contains(event.target as Node)) return;
      return callback(event);
    };
    document.addEventListener("focusin", listener);
    return () => {
      document.removeEventListener("focusin", listener);
    };
  }, [ref, callback]);
};
