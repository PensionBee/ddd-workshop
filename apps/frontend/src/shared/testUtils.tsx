import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
