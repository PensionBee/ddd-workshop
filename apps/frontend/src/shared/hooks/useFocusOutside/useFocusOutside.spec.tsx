import { act, fireEvent, render } from "@shared/testUtils";
import { useRef } from "react";
import { useFocusOutside } from "./useFocusOutside";

const MOCK_CALLBACK = jest.fn();

const FOCUS_EVENT = new MouseEvent("focusin", {
  bubbles: true,
  cancelable: true,
});

const Component = () => {
  const ref = useRef(null);
  useFocusOutside(ref, MOCK_CALLBACK);

  return (
    <div data-testid="parent-element">
      <div data-testid="ref-element" ref={ref}>
        <div data-testid="child-element" />
      </div>
      <div data-testid="outside-element" />
    </div>
  );
};

describe("useFocusOutside", () => {
  it("should fire callback when focusing outside the element", () => {
    const { asFragment, getByTestId } = render(<Component />);
    const outsideElement = getByTestId("outside-element");

    expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    act(() => fireEvent(outsideElement, FOCUS_EVENT));
    expect(MOCK_CALLBACK.mock.calls.length).toBe(1);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should fire callback when focusing on the elements parent", () => {
    const { asFragment, getByTestId } = render(<Component />);
    const parentElement = getByTestId("parent-element");

    expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    act(() => fireEvent(parentElement, FOCUS_EVENT));
    expect(MOCK_CALLBACK.mock.calls.length).toBe(1);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should NOT fire callback when focusing on the element", () => {
    const { asFragment, getByTestId } = render(<Component />);
    const refElement = getByTestId("ref-element");

    expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    act(() => fireEvent(refElement, FOCUS_EVENT));
    expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should NOT fire callback when focusing on the elements children", () => {
    const { asFragment, getByTestId } = render(<Component />);
    const childElement = getByTestId("child-element");

    expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    act(() => fireEvent(childElement, FOCUS_EVENT));
    expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    expect(asFragment()).toMatchSnapshot();
  });
});
