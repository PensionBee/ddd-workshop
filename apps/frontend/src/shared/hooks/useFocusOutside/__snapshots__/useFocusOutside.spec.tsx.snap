// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useFocusOutside > should fire callback when focusing outside the element 1`] = `
<DocumentFragment>
  <div
    data-testid="parent-element"
  >
    <div
      data-testid="ref-element"
    >
      <div
        data-testid="child-element"
      />
    </div>
    <div
      data-testid="outside-element"
    />
  </div>
</DocumentFragment>
`;

exports[`useFocusOutside should NOT fire callback when focusing on the element 1`] = `
<DocumentFragment>
  <div
    data-testid="parent-element"
  >
    <div
      data-testid="ref-element"
    >
      <div
        data-testid="child-element"
      />
    </div>
    <div
      data-testid="outside-element"
    />
  </div>
</DocumentFragment>
`;

exports[`useFocusOutside should NOT fire callback when focusing on the elements children 1`] = `
<DocumentFragment>
  <div
    data-testid="parent-element"
  >
    <div
      data-testid="ref-element"
    >
      <div
        data-testid="child-element"
      />
    </div>
    <div
      data-testid="outside-element"
    />
  </div>
</DocumentFragment>
`;

exports[`useFocusOutside should fire callback when focusing on the elements parent 1`] = `
<DocumentFragment>
  <div
    data-testid="parent-element"
  >
    <div
      data-testid="ref-element"
    >
      <div
        data-testid="child-element"
      />
    </div>
    <div
      data-testid="outside-element"
    />
  </div>
</DocumentFragment>
`;

exports[`useFocusOutside should fire callback when focusing outside the element 1`] = `
<DocumentFragment>
  <div
    data-testid="parent-element"
  >
    <div
      data-testid="ref-element"
    >
      <div
        data-testid="child-element"
      />
    </div>
    <div
      data-testid="outside-element"
    />
  </div>
</DocumentFragment>
`;
